### ISSUE TITLE
Request queue full on Publiclab.org

### ISSUE THREAD

**Issue by icarito** [publiclab/plots2#9716]
In the past 3 days we've had a recurring issue with the website responding with a 500 error and the logs showing "request queue full".
We've tripled the length of the request queue but still facing this issue every few hours,

**Comment by icarito** [comment]
Restarting the container fixes it momentarily.

**Comment by icarito** [comment]
This seems to be related to this error caught in Sentry:
```
Mysql2::Error::TimeoutError: Lock wait timeout exceeded; try restarting transaction: UPDATE `rusers` SET `last_request_at` = '2021-06-01 14:19:09', `updated_at` = '2021-06-01 14:19:09' WHERE `rusers`.`id` = xxx
```

**Comment by jywarren** [comment]
![image](https://user-images.githubusercontent.com/24359/120342192-8f181c80-c2c5-11eb-9b57-8b7fe106e6ba.png)

**Comment by anirudhprabhakaran3** [comment]
This seemed to help someone on StackOverflow, could help us too. https://chat.stackexchange.com/rooms/23589/discussion-between-javac-and-hitesh-mundra

**Comment by jywarren** [comment]
Happening across many routes actually:

![image](https://user-images.githubusercontent.com/24359/120344024-33e72980-c2c7-11eb-9cf8-85d944221d98.png)

So for example, these and others:

- https://publiclab.org/users/update
- https://publiclab.org/profile/________
- https://publiclab.org/login is getting hit 4x per minute

**Comment by icarito** [comment]
Increasing the process pool size to 20 from 16 seems to have resolved the request queue thus far.

**Comment by jywarren** [comment]
So we are suspecting that the timeout error and lock is the EFFECT and not the cause. When have peaks occurred, so we can try to trace an origin?

**Comment by jywarren** [comment]
Looking at one outage last night, it started right in this timeframe:

![image](https://user-images.githubusercontent.com/24359/120374531-a7009800-c2e7-11eb-9251-06221093af88.png)

UsersController#update was first to get really slow. Followed by about 10 other routes.

**Comment by anirudhprabhakaran3** [comment]
Sebastian, is there any way to find what exact transaction caused the failure?

**Comment by icarito** [comment]
Hi @anirudhprabhakaran3 thanks - we're working under the assumption that we're seeing the _effects_ of a surge in traffic to our omniauth routes which we've yet to explain.
Here's some interesting logs from `passenger.4000.log`:
```
App 249 output: I, [2021-06-02T20:19:25.267352 #249]  INFO -- omniauth: (google_oauth2) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:26.798838 #249]  INFO -- omniauth: (github) Request phase initiated.
App 379 output: I, [2021-06-02T20:19:28.118763 #379]  INFO -- omniauth: (facebook) Request phase initiated.
App 379 output: I, [2021-06-02T20:19:29.743321 #379]  INFO -- omniauth: (google_oauth2) Request phase initiated.
App 379 output: I, [2021-06-02T20:19:30.931253 #379]  INFO -- omniauth: (twitter) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:32.444230 #249]  INFO -- omniauth: (github) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:35.296289 #249]  INFO -- omniauth: (facebook) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:36.993363 #249]  INFO -- omniauth: (google_oauth2) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:37.928497 #249]  INFO -- omniauth: (facebook) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:40.305682 #249]  INFO -- omniauth: (twitter) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:53.223182 #249]  INFO -- omniauth: (twitter) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:56.407173 #249]  INFO -- omniauth: (google_oauth2) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:57.920310 #249]  INFO -- omniauth: (github) Request phase initiated.
App 249 output: I, [2021-06-02T20:19:59.552195 #249]  INFO -- omniauth: (facebook) Request phase initiated.
App 379 output: I, [2021-06-02T20:20:01.017987 #379]  INFO -- omniauth: (github) Request phase initiated.
```
We can see a lot of activity

**Comment by anirudhprabhakaran3** [comment]
Okay so this is my theory, don't know how much it might be correct.

There are multiple people trying to use third party authentication (like Google, GitHub as seen in the passenger logs). When that happens, the database is locked for some time, expecting that a response should come within that time. However, these auth operations are taking longer than the database is being locked for that particular thread, because of which we're getting a MySQL LockWaitTimeout error. 

This still doesn't explain the surge in the traffic but should be able to give us ideas about how to deal with this situation. One solution that I have seen on the internet is to increase the lock timeout. I'm not sure whether we have tied doing that. I'm also not aware of any other side effects that might occur because of the increased lock time.

**Comment by icarito** [comment]
The issue had to do with our storage volume hosted at GCS, we are using `gcsfuse` which seems to crash and lock the mountpoint, causing the processes to lock up and stop responding. The proposed solution is to implement Paperclip or use the Google Cloud storage API in other way to upload files https://github.com/publiclab/plots2/issues/9760

**Comment by icarito** [comment]
The underlying issue was described as well at https://github.com/publiclab/plots2/issues/9758 - closing this bug as the root cause has been found and a proposal for a permanent fix is at #9760
